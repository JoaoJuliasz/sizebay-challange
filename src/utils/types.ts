export type FilterType = 'done' | 'pending' | ''

export type Items = {
    value: string,
    status: 'done' | 'pending'
    itemIndex: number
}