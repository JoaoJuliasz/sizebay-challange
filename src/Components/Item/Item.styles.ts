import styled from "styled-components";
import { MinusOutlined, CheckOutlined } from '@ant-design/icons'

type ItemInputProps = {
    editCurrentItem: boolean
    done: boolean
}

type ButtonProps = {
    color: string
    hasBorder?: boolean
}

export const ItemContainer = styled.div`
    width: 80%;
    margin: 7px auto;

    @media only screen and (max-width: 600px) {
        width: 86%
    }
`

export const ItemComponentsContainer = styled.div`
    width: 100%;
    border: 1px solid #dbdbdb;
    border-radius: 5px;
    margin: auto;
`

export const ItemInput = styled.input<ItemInputProps>`
    width: 100%;
    max-width: ${(props) => props.editCurrentItem ? 'calc(100% - 100px)' : ''};
    padding: 12px 10px;
    outline: none;
    background: ${(props) => props.done ? '#00ff6263' : (!(props.editCurrentItem) ? '#f4f4f4' : '#fff')};
    border: none;
    border-radius: ${(props) => props.editCurrentItem ? '3px 0 0 3px' : '5px'};
    cursor: ${props => props.editCurrentItem ? '' : 'pointer'};
`

export const ItemButton = styled.button<ButtonProps>`
    padding: 13px 0;
    width: 50px;
    border: none;
    cursor: pointer;
    background: ${props => props.color};
    ${props => props.hasBorder && 'border-radius: 0 3px 3px 0'}
`

export const MinusIcon = styled(MinusOutlined) `
    background: #fff;
    padding: 4px;
    border-radius: 50%;
    color: #e34f4f;
`
export const CheckIcon = styled(CheckOutlined) `
    background: #fff;
    padding: 4px;
    border-radius: 50%;
    color: #68e291;
`
