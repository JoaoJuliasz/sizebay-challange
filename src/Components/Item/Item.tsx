import { useState, useEffect, useRef } from 'react'

//Importing libs
//@ts-ignore
import Tooltip from "reactjs-mappletooltip"
import { toast } from 'react-toastify';

import { Items } from '../../utils/types'

//Importing style-components 
import { ItemButton, ItemComponentsContainer, ItemContainer, ItemInput, MinusIcon, CheckIcon } from './Item.styles'

type ItemProps = {
    idx: number
    item: Items
    changeItemValue: (index: number, value: string) => void
    changeItemStatus: (index: number, action: string) => void
}

const Item = ({ idx, item, changeItemValue, changeItemStatus }: ItemProps) => {

    const [editCurrentItem, setEditCurrentItem] = useState<boolean>(false)

    const node = useRef<HTMLDivElement>(null)
    const inputNode = useRef<HTMLInputElement>(null)

    const handleClick = () => {
        setEditCurrentItem(true)
    }

    const isItemEmpty = () => !item.value

    const closeItemWhenItIsDone = () => {
        if (!isItemEmpty()) {
            changeItemStatus(item.itemIndex, 'finish')
            setTimeout(() => {
                setEditCurrentItem(!editCurrentItem)
            }, 0)
        }
    }

    const handleClickOutside = ({ target }: MouseEvent | TouchEvent) => {
        if (node?.current?.contains(target as Node)) {
            return
        }
        if (isItemEmpty()) {
            toast.error("Current item can't be empty")
            inputNode?.current?.focus()
            return
        }
        setEditCurrentItem(false)
    }

    useEffect(() => {
        if (editCurrentItem) {
            document.addEventListener("mousedown", handleClickOutside)
        } else {
            document.removeEventListener("mousedown", handleClickOutside)
        }

        return () => {
            document.removeEventListener("mousedown", handleClickOutside)
        }
    }, [editCurrentItem])

    return (
        <ItemContainer key={idx} id="item-container"
            onClick={() => handleClick()}
            ref={node}
        >
            <Tooltip style={{ width: '100%' }} direction={'bottom'} showMappleIf={!editCurrentItem}>
                <ItemComponentsContainer>
                    <ItemInput id="item-input" ref={inputNode}
                        value={item.value} type="text"
                        onChange={e => changeItemValue(item.itemIndex, e.target.value)}
                        editCurrentItem={editCurrentItem} done={item.status === 'done'} />
                    {editCurrentItem &&
                        <>
                            <ItemButton id="remove-btn" color={'#e34f4f'} onClick={() => changeItemStatus(item.itemIndex, 'remove')}>
                                <MinusIcon />
                            </ItemButton>
                            {<ItemButton id="done-btn" color={'#68e291'} hasBorder={true} onClick={() => closeItemWhenItIsDone()}>
                                <CheckIcon />
                            </ItemButton>}
                        </>
                    }
                </ItemComponentsContainer>
                <div>
                    Edit task
                </div>
            </Tooltip>
        </ItemContainer>
    );
};

export default Item;