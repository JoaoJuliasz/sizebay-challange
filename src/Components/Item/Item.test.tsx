import { shallow } from 'enzyme'
import Item from './Item'
import { Items } from '../../utils/types';

describe('Item', () => {

    const mockIndex = 0
    const mockItem:Items = {value: 'item 1', status: 'done', itemIndex: 0}
    const mockChangeItemValue = jest.fn()
    const mockChangeItemStatus= jest.fn()

    const itemWrapper = shallow(<Item idx={mockIndex} item={mockItem} changeItemStatus={mockChangeItemStatus} changeItemValue={mockChangeItemValue} />)

    it('expect to render Item component', () => {
        expect(itemWrapper).toMatchSnapshot()
    })

    it('expect to item buttons should be displayed', () => {
        const itemContainer = itemWrapper.find('[id="item-container"]')
        itemContainer.simulate('click')
        expect(itemWrapper.find('[id="remove-btn"]').length).toBe(1)
        expect(itemWrapper.find('[id="done-btn"]').length).toBe(1)
    })

    it('expect to item input value should change', () => {
        const itemInput = itemWrapper.find('[id="item-input"]')
        itemInput.simulate('change', {target: {value: {index: 0, value: 'item 2'}}})
        expect(mockChangeItemValue).toBeCalled()
    })

    it('expect to item status should be "done"', () => {
        const itemContainer = itemWrapper.find('[id="item-container"]')
        itemContainer.simulate('click')
        itemWrapper.find('[id="done-btn"]').simulate('click')
        expect(mockChangeItemStatus).toBeCalled()
    })

    it('expect to item value can not be empty', () => {
        const itemInput = itemWrapper.find('[id="item-input"]')
        itemInput.simulate('change', {target: {value: {index: 0, value: ''}}})
        setTimeout(() => {
            itemWrapper.find('[id="done-btn"]').simulate('click')
            expect(mockChangeItemStatus).not.toBeCalled()
        }, 500)
    })
})