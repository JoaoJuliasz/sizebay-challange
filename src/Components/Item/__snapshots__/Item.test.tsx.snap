// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Should render Item component 1`] = `ShallowWrapper {}`;

exports[`Item expect to render Item component 1`] = `ShallowWrapper {}`;
